### PR TITLE
Add delivery type to PerformanceResourceTiming.

### PR DESCRIPTION
**Pull request by jeremyroman**
Addresses #332.

This defines a concept of "delivery type" which currently describes the difference between resources delivered from the HTTP cache and those not (currently only detectable via transferSize). It is expected to be extended to describe resources delivered from other stores, such as the consuming resources from the HTML preload buffer and navigational prefetches.

***
<a href="https://pr-preview.s3.amazonaws.com/jeremyroman/resource-timing/pull/343.html" title="Last updated on Mar 10, 2023, 4:06 PM UTC (78dc701)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/resource-timing/343/98b0e13...jeremyroman:78dc701.html" title="Last updated on Mar 10, 2023, 4:06 PM UTC (78dc701)">Diff</a>

### CODE REVIEW

**Comment by jeremyroman** [comment]
This is missing WPT tests at present. I understand if merging this is blocked on that, but would you mind having a look @noamr ? I expect the tests for this will be relatively straightforward (at least, once we have an implementation in Chromium to avoid test bugs).

File: index.html
Comment (yoavweiss) [inline index.html:1148]: s/Tode/Type/ here and below

Comment (jeremyroman) [reply index.html]: Done.

File: index.html
Comment (yoavweiss) [inline index.html:1148]: deliveryType

**Review by yoavweiss — APPROVED** [review]
LGTM, but I'd love to run this by the WG and see if there are any objections, as well as get tests in place before this lands.

**Comment by jeremyroman** [comment]
> LGTM, but I'd love to run this by the WG and see if there are any objections, as well as get tests in place before this lands.

Is TPAC a good venue for that? If so are there steps I should take to get it on the agenda?

**Comment by yoavweiss** [comment]
> Is TPAC a good venue for that? If so are there steps I should take to get it on the agenda?

We can probably cover this on [today's call](https://docs.google.com/document/d/10dz_7QM5XCNsGeI63R864lF9gFqlqQD37B4q8Q46LMM/edit?pli=1#heading=h.15anydpwx5wv).

**Comment by nicjansma** [comment]
Summary from W3C WebPerf call:

* Need to figure out cross-origin/TAO/CORS limitations
* Possible objections from Safari, to be discussed further at TPAC in ~2 weeks

**Comment by yoavweiss** [comment]
Thinking out loud regarding the need for restrictions here:
If we don't add any TAO restrictions to this, this doesn't reveal anything about the resource itself, but it can reveal some metadata about the resource. Specifically, it can reveal things about the resource's caching headers. It provides a bit on whether the resource is cachable or not (and potentially, for how long it is cacheable, through continous probing).

While not significant on its own, this could be sufficient to reveal user state in cases where the cachability of the resource varies between e.g. logged-in and non-logged-in users. 

So, I think we want this to be TAO protected.

/cc @arturjanc and @mikewest to tell me if I'm being overly paranoid about this or overly optimistic about this not already being leaked through timing.

**Comment by jeremyroman** [comment]
I think that tracks. I think the strongest counterarguments would be:

1. The risk is substantially mitigated in browsers which partition the cache by at least top-level site (and some also use a secondary key). In such cases the cache state doesn't reveal anything about browsing activity on other top-level sites. I'm not certain but I believe this is now all major browsers.
2. It is likely already deducible from timing heuristics (even without TAO-protected info) in most cases.

I'm not opposed to TAO protection, though, and presumably anyone gathering RUM is also serving their important cross-origin resources with a suitable TAO header.

If that is the resolution, the behavior I propose is that the if `"cache"` would have been the delivery type but the response's [timing allow passed flag](https://fetch.spec.whatwg.org/#concept-response-timing-allow-passed) is unset, then it returns `""` instead. This is probably slightly conservative since it would protect it if a redirect didn't pass the TAO check even if the final resource did, but similar concerns might apply to the destination of the redirect and there's value in being consistent with other TAO protected properties.

**Comment by jeremyroman** [comment]
On further investigation, I think the current text in this PR already implies a TAO check in the following way:

Fetch says: 
> If response’s [timing allow passed flag](https://fetch.spec.whatwg.org/#concept-response-timing-allow-passed) is not set, then set timingInfo to a the result of [creating an opaque timing info](https://fetch.spec.whatwg.org/#create-an-opaque-timing-info) for timingInfo, set bodyInfo to a new [response body info](https://fetch.spec.whatwg.org/#response-body-info), and set cacheState to the empty string.

This is then used to populate the cache mode and thus the default delivery type, so "setup the resource timing entry" won't see the `"cache"` delivery type if the timing allow passed flag wasn't set (i.e., the response didn't pass a TAO check).

**Comment by jeremyroman** [comment]
Status update:
* the question of the TAO is, I believe, resolved above
* re. TPAC discussion, user agents which restrict access to certain properties beyond the specified TAO checks can, of course, apply similar restrictions to this property if applicable; if explicit verbiage to permit this lands in the specification then I think it will generalize naturally
* slated to go to origin trial in Chrome 110 (including an extension for `"navigational-prefetch"`, which is otherwise difficult to detect)
* WPT coverage is landed as tentative ([resource-timing/delivery-type.tentative.any.js](https://github.com/web-platform-tests/wpt/blob/master/resource-timing/delivery-type.tentative.any.js))

I believe that covers the identified items. @yoavweiss what is the next step under the group's working mode?

**Comment by yoavweiss** [comment]
^^ @achristensen07 @bdekoz

**Comment by jeremyroman** [comment]
#369 seems to be general verbiage that I think covers any application it would have to this PR. I'd like to renew my request to know what the next step/milestone here is.

**Comment by noamr** [comment]
This is LGTM, @yoavweiss do you want to take another look?

**Review by yoavweiss — APPROVED** [review]
LGTM

**Pull request merged**
